### DISCUSSION TITLE
decrease quantum for osu!lazer when running with linux and pipewire?

### DISCUSSION THREAD

**Discussion by fxzzi** [ppy/osu#30527] [Q&A] [answered]
Hi all, I've attempted what I could to decrease latency for my system and for osu!lazer specifically, I followed this guide:

https://www.reddit.com/r/linux_gaming/comments/1gao420/low_latency_guide_for_linux_using_pipewire/

After setting everything up, it seems osu!lazer (alsa) is still limited to a quantum of 441, whereas my min and max are 64 and 128 respectively. How could I get osu! to respect my max quantum and therefore lower my audio latency further?

**Comment by sineplusx** [comment, marked as answer]
Lazer currently has a hardcoded period size ("quantum") of 10ms at a 44100Hz output sample rate. You can force the period/buffer lower with environment variables when running lazer:

```
PIPEWIRE_LATENCY=64/44100
PIPEWIRE_ALSA='{ alsa.format=S32_LE alsa.channels=2 alsa.buffer-bytes=512 alsa.period-bytes=256 }'
```

This question has already been asked before, but it seems the original thread got taken down for some reason.

**Reply by bdach** [reply]
We don't "take down" threads, so users that posted the previous discussions likely either "closed" their discussion threads which is a thing you can do now and which shadow-realms the discussions out of search results, or they deleted their accounts, which also does that but harder.

**Reply by fxzzi** [reply]
Thanks! using just the top env var seems to work fine and pw-top shows 64 perfectly. However using the second one as well causes my osu to be slowed down and crackly.

**Reply by Irgendwer01** [reply]
@fxzzi @sineplusx I know this is an old Thread, but the reason for being so slow seems to be `alsa.buffer-bytes` being set too low, setting it to 1024 fixes it.

**Reply by sineplusx** [reply]
How low you can set it depends on your system and config. These specific values are only an example that happened to work for me.

**Comment by rszyma** [comment]
(I wanted to this in the original thread but can't find it either, so I'll share it here.)

I found another (better?) way to decrease latency on Linux -- by bypassing Pipewire completely. 

Note, that by doing this you temporarily lose system audio, so audio from discord / chrome / etc will not work until re-enabled.

How to:
1. Stop osu and any other app that might play audio
2. Run `systemctl stop --user pipewire.service pipewire.socket pipewire-pulse.service pipewire-pulse.socket`
3. In osu in-game ssettings there's audio device selector, select the name of your physical output device.
4. Done. Now you should be running exclusive audio.
5. After being done with playing osu, close it and run `systemctl start --user pipewire.service pipewire.socket pipewire-pulse.service pipewire-pulse.socket` to restore system audio. Or simply rebooting would work too.

From my experience I think it actually works better than tweaking `PIPEWIRE_LATENCY` or whatever other pipewire settings (and I've been tweaking pipewire for the last 2 years on and off). Without pipewire I just got the best PP score in 6 years after 10 minute session, so it actually seems to help and not being another placebo xD. But idk, might be just skills coming back to my hands.

**Reply by rszyma** [reply]
actually no need for stopping all the services in step 2. - as long as you close all apps that hold audio stream it should be good. You can check for such apps with `pw-top`.

**Reply by rszyma** [reply]
Update: it's possible to go even lower latency. Currently there's hardcoded 20ms latency for audio device buffer. With [this patch](https://github.com/ppy/osu-framework/pull/6724), in numbers, I got it down to ~4ms (176/44100*1000ms).

**Reply by sineplusx** [reply]
I have also been using exclusive ALSA for the last few months, it works really well.

Adding control over the BASS variables has been tried once (shot down) and suggested many times before.

https://github.com/ppy/osu-framework/pull/4008
https://github.com/ppy/osu/discussions/27058

If you want to have the lowered latency on something other than a dev build, you can try my diff from https://github.com/ppy/osu-framework/issues/6647
